File: apps/api/internal/route/handshake.go
package route

import (
	"encoding/json"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/lapanxd/volatus-api/internal/dto"
	"github.com/lapanxd/volatus-api/internal/model"
	"github.com/lapanxd/volatus-api/internal/service"
	"github.com/lapanxd/volatus-api/internal/sse"
	"gorm.io/gorm"
)

func HandshakeRoutes(r *gin.RouterGroup, db *gorm.DB) {
	r.POST("/init", func(c *gin.Context) {
		HandshakeInit(c, db)
	})

	r.POST("/response", func(c *gin.Context) {
		HandshakeResponse(c, db)
	})

	r.GET("/pending", GetPendingHandshake)
}

func HandshakeInit(c *gin.Context, db *gorm.DB) {
	var req dto.InitInput
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	fromUserID := c.GetUint("user_id")
	if !service.CheckIfUserExists(db, req.ToUserID) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "target user does not exist"})
		return
	}

	sessionID, err := service.InitHandshake(fromUserID, req.ToUserID, req.SDPOffer)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, dto.InitOutput{SessionID: sessionID})
}

func HandshakeResponse(c *gin.Context, db *gorm.DB) {
	var req dto.ResponseInput
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid input"})
		return
	}

	userID := c.GetUint("user_id")
	session, err := service.RespondHandshake(userID, req.SessionID, req.Accepted)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if session == nil {
		// return 200 even if handshake is not accepted
		c.Status(http.StatusOK)
		return
	}

	succeeded := sse.SendEvent(session.FromUserID, sse.SSEEvent{
		EventType: "handshake_accepted",
		Payload: dto.AcceptedPayloadOutput{
			FromUserID: session.FromUserID,
			ToUserID:   session.ToUserID,
			SDPAnswer:  req.SDPAnswer,
		},
	})

	if !succeeded {
		payloadBytes, err := json.Marshal(dto.AcceptedPayloadOutput{
			FromUserID: session.FromUserID,
			ToUserID:   session.ToUserID,
			SDPAnswer:  req.SDPAnswer,
		})
		if err != nil {
			c.Status(http.StatusOK)
			return
		}

		notification := model.PendingNotification{
			UserID:    session.FromUserID,
			EventType: "handshake_accepted",
			Payload:   string(payloadBytes),
		}

		db.Create(&notification)
	}

	c.Status(http.StatusOK)
}

func GetPendingHandshake(c *gin.Context) {
	userID := c.GetUint("user_id")
	pending := service.GetPendingHandshakes(userID)
	c.JSON(http.StatusOK, pending)
}

File: apps/client/src-tauri/src/commands/handshake.rs
use crate::api::dtos::handshake::PendingHandshakeOutputDto;
use crate::{API_CLIENT, API_URL};

#[tauri::command]
pub async fn handshake_init(to_user_id: u32, sdp_offer: String) -> Result<(), String> {
    let res = API_CLIENT
        .post(
            &format!("{}/handshake/init", *API_URL),
            serde_json::json!({ "to_user_id": to_user_id, "sdp_offer": sdp_offer }),
        )
        .await
        .map_err(|e| e.to_string())?;

    if res.status().is_success() {
        Ok(())
    } else {
        Err("Handshake init failed".into())
    }
}

#[tauri::command]
pub async fn handshake_response(
    session_id: String,
    accepted: bool,
    sdp_answer: String,
) -> Result<(), String> {
    let res = API_CLIENT
        .post(
            &format!("{}/handshake/response", *API_URL),
            serde_json::json!({ "session_id": session_id, "accepted": accepted, "sdp_answer": sdp_answer }),
        )
        .await
        .map_err(|e| e.to_string())?;

    if res.status().is_success() {
        // start p2p communication
        Ok(())
    } else {
        Err("Handshake response failed".into())
    }
}

#[tauri::command]
pub async fn get_pending_handshakes() -> Result<Vec<PendingHandshakeOutputDto>, String> {
    let res = API_CLIENT
        .get(&format!("{}/handshake/pending", *API_URL))
        .await
        .map_err(|e| e.to_string())?;

    if res.status().is_success() {
        let body: Vec<PendingHandshakeOutputDto> = res.json().await.map_err(|e| e.to_string())?;
        Ok(body)
    } else {
        Err("Handshake response failed".into())
    }
}
